docs/chymyst07.md

```diff
@@ -783,10 +793,173 @@ go { case barrier1(x1, reply1) + barrier2(x2, reply2) => ... }
+It is clear that `replyB()` should come last: this is the reply to the `barrier()` molecule, which should be performed only after we counted up to `n`.
+Until then, this reaction must be blocked.
+
+The only way to block in the middle of a reaction is to emit a blocking molecule there.
+Therefore, some blocking molecule must be emitted in the reaction before replying to `barrier()`.
+Let us make `counter()` a blocking molecule:
+
+```scala
+go { case barrier(_, replyB) + counter(k, replyC) => ???; if (k+1<n) counter(k+1); ???; replyB() }
+
+```
+
+What is still missing is the reply to the `counter(k)` molecule.
+Now we are faced with a question: at what point should we perform that reply, before emitting `counter(k+1)` or after?
+We could write one of the two reactions:
+
+```scala
+val reaction1 = go { case barrier(_, replyB) + counter(k, replyC) => replyC(); if (k+1<n) counter(k+1); replyB() }
+val reaction2 = go { case barrier(_, replyB) + counter(k, replyC) => if (k+1<n) counter(k+1); replyC(); replyB() }
+
+```
+
+We now need to decide whether `reaction1` or `reaction2` works as we want.
+
+To resolve this question, let us visualize the molecules that we'd like to be present at different stages of running the program.
+Suppose that `n=10` and we have already consumed all `barrier` molecules except three.
+We presently have three `barrier()` molecules and one `counter(7)` molecule (which, as we decided, is a blocking molecule).
+
+Note that the `counter(7)` molecule was emitted by a previous reaction of the same kind, 
+
+`barrier() + counter(6) => ... counter(7); ... }`
+
+So, this previous reaction is now blocked at the place where it emitted `counter(7)`. 
+
+Next, the reaction `barrier() + counter(7) => ...` will start and consume its input molecules, leaving two `barrier()` molecules present.
+
+Now the reaction body of `barrier() + counter(7) => ...` will run and emit `counter(8)`.
+Then we will have the molecules `barrier(), barrier(), counter(8)` in the soup.
+This set of molecules will be the same whether we define `reaction1` or `reaction2`.
+However, in `reaction1` the reply is already sent to `counter(7)` before emitting `counter(8)`.
+This will unblock the previous reaction,
+ 
+```scala
+{ case barrier(_, replyB) + counter(6, replyC) => replyC(); counter(7); replyB() }
+
+```
+
+which was blocked at the call to `counter(7)`.
+Now this reaction can proceed to reply to `barrier()` using `replyB()`.
+However, at this point it is too early to send a reply to `barrier()`, because we still have two `barrier()` molecules that we have not yet consumed!
+
+Therefore, `reaction1` is incorrect.
+On the other hand, `reaction2` works correctly.
+It will block at each emission of `counter(k+1)` and unblock only when `k=n-1`.
+At that time, `reaction2` will reply to the `counter(n-1)` molecule and to the last consumed `barrier()` molecule.
```

Comment:
That's not the same thing. The `barrier()` molecule last emitted is not necessarily the one last consumed. Many molecules can be emitted but then reactions will start in an arbitrary order. But we reply to the `barrier()` molecule only after it's consumed by some reaction.